src/webui/api/torrentscontroller.cpp

```diff
@@ -873,15 +873,47 @@ void TorrentsController::editTrackerAction()
 
 void TorrentsController::removeTrackersAction()
 {
-    requireParams({u"hash"_s, u"urls"_s});
+    requireParams({u"urls"_s});
 
-    const auto id = BitTorrent::TorrentID::fromString(params()[u"hash"_s]);
-    BitTorrent::Torrent *const torrent = BitTorrent::Session::instance()->getTorrent(id);
-    if (!torrent)
-        throw APIError(APIErrorType::NotFound);
+    const QString hashParam = params()[u"hash"_s];
+    const QStringList urlsParam = params()[u"urls"_s].split(u'|', Qt::SkipEmptyParts);
 
-    const QStringList urls = params()[u"urls"_s].split(u'|');
-    torrent->removeTrackers(urls);
+    // remove trackers from specified torrent
+    if (!hashParam.isEmpty())
+    {
+        const auto id = BitTorrent::TorrentID::fromString(hashParam);
+        BitTorrent::Torrent *const torrent = BitTorrent::Session::instance()->getTorrent(id);
+        if (!torrent)
+            throw APIError(APIErrorType::NotFound);
+
+        torrent->removeTrackers(urlsParam);
+    }
+    // remove trackers from all torrents
+    else
+    {
+        QStringList urls;
+        urls.reserve(urlsParam.size());
+        for (const QString &urlStr : urlsParam)
+            urls << QUrl::fromPercentEncoding(urlStr.toUtf8());
+
+        for (BitTorrent::Torrent *const torrent : asConst(BitTorrent::Session::instance()->torrents()))
+        {
+            const QList<BitTorrent::TrackerEntryStatus> trackers = torrent->trackers();
+            QStringList trackersToRemove;
+            trackersToRemove.reserve(trackers.size());
+            for (const BitTorrent::TrackerEntryStatus &trackerEntryStatus : trackers)
+            {
+                for (const QString &url : urls)
+                {
+                    if ((trackerEntryStatus.url == url) || (QUrl(trackerEntryStatus.url).host() == url))
+                        trackersToRemove.append(trackerEntryStatus.url);
```

Comment:
Why you need to reinvent the wheel instead of using `torrent->removeTrackers()`?
I imagine the difference between "remove trackers from specified torrent" and "from all torrents" would be only the torrent range. And they all delegate the work to `torrent->removeTrackers()`.

Reply:
See `TrackersFilterWidget::removeTracker` for an example of why we need this logic. The tracker shown in sidebar is just the hostname, not the full URL.

Reply:
>See TrackersFilterWidget::removeTracker for an example of why we need this logic. The tracker shown in sidebar is just the hostname, not the full URL.

Does that mean `urls` has different format/meaning depending on `hash` value? If so, it is bad API design.

Reply:
> See `TrackersFilterWidget::removeTracker` for an example of why we need this logic. The tracker shown in sidebar is just the hostname, not the full URL.

It is logic of UI layer. It is UI layer that reduces the real addresses of trackers, so it must take care of it itself. The simplest option is to store a list of real addresses related to each displayed one.

Reply:
IMO, this behaviour (i.e. remove partially matched trackers) is acceptable at API level if both of the following conditions are met:
1. it works in the same way independently of torrents it is applied to (whether single, multiple or all of them),
2. this behaviour makes sense by itself (without regards to some particular UI).

Reply:
That's part of the reason I originally had this as its own unique API endpoint. Happy to separate the logic back out again.

Reply:
> That's part of the reason I originally had this as its own unique API endpoint. Happy to separate the logic back out again.

It doesn't change the point. They still remain essentially identical actions, differing only in affecting one or another number of torrents. So it would still look inconsistent if they differed in something else. 
The following question remains open:
Whether "this behaviour makes sense by itself (without regards to some particular UI)"?
If the answer is Yes, then it makes sense to behave like this (to allow partial matching of tracker addresses) in both cases under discussion (one or all torrents).
If not, then it is more appropriate to transfer this logic to the layer that requires it. https://github.com/qbittorrent/qBittorrent/pull/21056#discussion_r1706600719

Reply:
Ah I see your point here. Will switch to requiring the full URL.